商品情報編集機能

ken-forest opened the pull request:
# What
出品した商品情報を編集する機能を作る

# Why
もし、出品者が品物を出した時点で載せている情報が間違っていた場合、早急に訂正する必要が出てくる。そういう時のための必須機能である。

＜Gyazoで撮影したもの＞
・ログイン状態の出品者は、商品情報編集ページに遷移できる
https://gyazo.com/3a85f847cbc96cd8d69e548a3eaa230f

・必要な情報を適切に入力して「更新する」ボタンを押すと、商品の情報を編集できる
https://gyazo.com/3bf4b17bedaad68f6905b1f27ae08bb2

・入力に問題がある状態で「変更する」ボタンが押された場合、情報は保存されず、編集ページに戻りエラーメッセージが表示される
https://gyazo.com/dae1b12be841ac504ed5f70b056e2ede

・何も編集せずに「更新する」ボタンを押しても、画像無しの商品にならない
https://gyazo.com/f58d28b2a079eedfc470248f9279a642

・ログイン状態の場合でも、URLを直接入力して自身が出品していない商品の商品情報編集ページへ遷移しようとすると、商品の販売状況に関わらずトップページに遷移する
①ユーザー「ken」のオムライスの絵画、の編集ページURL
https://gyazo.com/e48b673b8b401ff2ce7aadeebc571e25

②そこへ侵入しようとするユーザー「mori」
https://gyazo.com/ac32a827c6b5867cf0de6a9b5955e328

・ログアウト状態の場合は、URLを直接入力して商品情報編集ページへ遷移しようとすると、商品の販売状況に関わらずログインページに遷移する
https://gyazo.com/6c9e50ce3fb09408e5d59e26c8682875

・商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される（商品画像・販売手数料・販売利益に関しては、表示されない状態で良い）
https://gyazo.com/6a38c364da44101145c17b8b874de027

ken-forest committed:
itemsコントローラにeditアクションとupdateアクションを定義

ken-forest committed:
詳細ページの編集ボタンに編集ページへのパスを記述

app/views/items/show.html.erbの39行目のlink_toのパスに、
edit_item_path(@item.id)
を記述

ken-forest committed:
itemsコントローラのeditアクションに、インスタンス変数@itemを定義

ken-forest committed:
編集画面上部のform_withにeditアクションの@itemを渡した

ken-forest committed:
app/views/items/edit.html.erbの全てのhogeに、該当する情報を記述

ken-forest committed:
itemsコントローラのupdateアクションの中身を記述

app/controllers/items_controller.rbに記述を追加
・インスタンス変数@itemを定義
・編集成功した後は、商品詳細画面にリダイレクトする
・編集が失敗時、そのまま編集画面に戻される

ken-forest committed:
編集失敗時、エラーメッセージが出るようにした

ken-forest committed:
ログイン中のユーザでも、他のユーザのeditページへ飛べないようにした

ken-forest commented:
レビューをよろしくおねがいします！

strongmasahiro commented on app/controllers/items_controller.rb:
```diff
@@ -25,11 +26,35 @@ def show
     @item = Item.find(params[:id])
   end
 
+  def edit
+    @item = Item.find(params[:id])
+  end
+
+  def update
+    @item = Item.find(params[:id])
+    if @item.update(item_params)
+      redirect_to item_path(@item.id)
+    else
+      render :edit
+    end
+  end
```
```
@item = Item.find(params[:id])
```
こちらはshow,edit,update内で定義するのではなく、private以下にメソッド化して、beforeアクションで呼び出す形に修正しておきましょう。
複数回使用する変数をメソッド化することで可読性が上がり、改修作業が発生した際も手を加えやすくなります。

strongmasahiro reviewed (commented):
レビュー依頼ありがとうございます！
すごく綺麗に記述されていますね！
コメントしましたので、ご確認をお願いいたします！

ken-forest committed:
「@item = Item.find(params[:id])」をメソッドとして定義

itemsコントローラにて、show・edit・updateアクション内にあった、
「@item = Item.find(params[:id])」
、、をdefine_itemメソッドとして定義して、before_actionとして発動するように記述した

ken-forest commented:
ご指摘いただいた、１箇所を修正しました。
それぞれのアクションで挙動は問題ありませんでした。
引き続き、レビューをよろしくお願いします！

yuhi-di commented on app/controllers/items_controller.rb:
```diff
@@ -22,14 +23,42 @@ def create
 
+  def move_to_top_page
+    @item = Item.find(params[:id])
+
+    unless user_signed_in? 
+      redirect_to root_path 
+    end
```
**こちらは任意修正なりますのでフォーマットの提出は不要となります！**
より良いコードを書くための知識としてご確認いただけますと幸いです。

以下の記述を削除しましょう。

```
unless user_signed_in? 
      redirect_to root_path 
end
```

理由
・authenticate_user!と記述が重複するため。

yuhi-di reviewed (commented):
修正とレビュー依頼ありがとうございます。

しっかりと実装できていることが確認できましたので、
**LGTM**とさせていただきます！

１点、任意修正箇所を挙げさせていただきました。
フォーマットの提出は必要ございません。ご確認していただけますと幸いです。

商品編集機能の実装、お疲れ様でした。
引き続き、応援しております。

ken-forest committed:
アドバイスをいただいた箇所をコメントアウトしました
